### ISSUE TITLE
Getting Nan after the first batch 

### ISSUE THREAD

**Issue by QLgogo** [Michaelvll/DeepCCA#4]
After the first batch, I got all NANs after nn.Linear() given normal inputs. I check the gradients in the layers which have the NANs after the update by print(model.layer.weight.grad) and found the gradients are all NANs. Have you met this problem before?

**Comment by Michaelvll** [comment]
Yes! The deep CCA maybe somewhat numerical unstable because of the decomposition operations for calculating CCA loss. I did struggle with it when I tried to use it in my another project (and eventually gave up). It may work if you change the float32 to float64.

**Comment by QLgogo** [comment]
Thanks for your reply. Changing float32 to float64 doesn't work for my dataset. Do you have any other suggestion?

**Comment by Michaelvll** [comment]
Sorry, but I do not have any more suggestion for it. I am wondering if you use a smaller output size may improve the numerical stability since only the largest ``output_size`` singular values will be used for calculation when you turn ``use_all_singular_values`` off. And maybe you can, if you would like to, take a look at objectives.py and uncomment those assertion for checking NANs, and find out what causes them. I will be very appreciated if you can figure it out and be so kind to let me know about how to fix it.
B.t.w. the ``Run.ipynb`` is an example for this dcca method on mnist dataset that works. If you don't mind taking a look at those settings there, it may help you.

**Comment by QLgogo** [comment]
Sorry I cannot figure it out either. I have turned to use Matlab version of DCCA. Matlab seems more stable and doesn't have this nan problem.

**Comment by arnaghosh** [comment]
Hi,
I am facing a similar issue with the deep CCA. It seems the Nan comes because the function 'SymeigBackward' returns Nan values!
I tried playing a bit with the regularization parameters and it seems if I reduce the r1 to 1e-6 (instead of 1e-3), it works. Note that I was facing the Nan issue for SigmaHat11 and hence I was tweaking r1. Although I am not sure why this happens from the point of the mathematics. Any help would be much appreciated. 
Thanks

**Comment by xuanhanyu** [comment]
I see that there are four version, i.e., pytorch,keras, matable, C++. What are the differences between these versions?

**Comment by Michaelvll** [comment]
@qpmnh Hi, the pytorch version is directly adapted from the [keras version](https://github.com/VahidooX/DeepCCA). Sorry, but I have no idea what is the difference to the matlab and C++ version. Since backward support for calculating eigenvalue is just added to pytorch from 1.0 version, there may be some instability, which gives NAN.

**Comment by xuanhanyu** [comment]
Cause of Nan may be the original data of mini-batch is all of 0. I tried  to set batch_size to 50000, rather than 800. The problem of Nan is solved.

**Comment by arminarj** [comment]
Hi dear all,
I have tried multiple output sizes for my data, but output_size=2 (input_size=300) only worked for me, which was not good at all.
However, Duo to new #5 , I have tried various output sizes (up to the input_dim) for my dataset for a couple of epochs and worked completely fine for me.
Dear @Michaelvll , I believe you can close this issue for now.
Please let me know the case you still have a problem with Nan gradients.
Best,
Armin

**Comment by Michaelvll** [comment]
> Hi dear all,
> I have tried multiple output sizes for my data, but output_size=2 (input_size=300) only worked for me, which was not good at all.
> However, Duo to new #5 , I have tried various output sizes (up to the input_dim) for my dataset for a couple of epochs and worked completely fine for me.
> Dear @Michaelvll , I believe you can close this issue for now.
> Please let me know the case you still have a problem with Nan gradients.
> Best,
> Armin

Thank you very much for your work! I really appreciate your contribution. The solution looks good to me. I merged your PR to the master branch and reran the Run.ipynb with the new codes. I will close this issue by now. Please feel free to have further discussions here, and let me know if we should reopen this issue.

**Comment by jameschapman19** [comment]
Hi Michael et al. - thanks for your work here. I'm migrating to pytorch from a tensorflow implementation. An observation I have made is that lots of the tensorflow implementations use trace(sqrt(T'T)) where you use sqrt(trace(T'T). My understanding (and I've manually checked the correlations) is that the former produces the sum of correlations and the latter produces something else (it's the difference between sqrt(a+b) and sqrt(a)+sqrt(b)). If I edit your work to the sum of correlations way pytorch gradients break because T'T has negative entries.

My theory isn't quite good enough to answer the question of whether T'T should have negative entries but given that it works fine in tf it looks like it probably relates to the eigenvectors produced by symeig compared to tf.eigh

**Comment by jameschapman19** [comment]
I've put on this thread because the change results in Nans after first batch (iff T'T has negative entries).

**Comment by Michaelvll** [comment]
Hi @jameschapman19,
Thank you for your comment! I am not sure which tf codebase uses trace(sqrt(T'T)), but according to the equation (10) in the original DeepCCA paper, the implementation in our code might be correct. Also, in the tf edition [VanhidooX/DeepCCA](https://github.com/VahidooX/DeepCCA), they use sqrt(trace(T'T)). The diagonal of T'T is the norm of the row vectors in T, so the trace of T'T is non-negative.

**Comment by jameschapman19** [comment]
See e.g. https://bitbucket.org/qingming_tang/deep-canonical-correlation-analysis/src/master/CCA.py

Also in the paper just below (13) they write something like for the gradient we require the trace of the sqrt of T'T which is a bit more clear about order of operations. 

I'm playing around myself if I find the solution in pytorch though I fear it's to do with the symeig not working in the same way as eigh

**Comment by jameschapman19** [comment]
Actually if you look at your branch for not using all singular values or the tensorflow one you quoted, you'll see they are of the form sum(sqrt(T'T))

**Comment by jameschapman19** [comment]
Really should have raised a new issue but not commented on anything open source before sorry!

**Comment by Michaelvll** [comment]
Thank you for pointing out! I think you are right. Please open a new issue if you would like to!

**Comment by Michaelvll** [comment]
Hi @jameschapman19,
Thank you for your suggestion! I fixed the order of the trace and sqrt you mentioned in the latest commit `2fa648e4`. Please feel free to open a new issue if you have any further concerns! :)

**Comment by vandana-rajan** [comment]
Hi,

I tried all the suggestions and still I was getting "Function 'SymeigBackward' returned nan values in its 0th output." error. So I modified the code like this,

```
[D1, V1] = torch.symeig(SigmaHat11+eps*torch.rand([SigmaHat11.shape[0],SigmaHat11.shape[1]]).to(self.device), eigenvectors=True)
[D2, V2] = torch.symeig(SigmaHat22+eps*torch.rand([SigmaHat22.shape[0],SigmaHat22.shape[1]]).to(self.device), eigenvectors=True)
```

Now I don't get the error anymore. So, my question is, is it okay to add a random perturbation matrix to the covariance matrix before inputting to the symeig function? Is it theoretically or programmatically correct?

Thanks,

**Comment by arminarj** [comment]
Hi @vandana-rajan, 

Duo to the small values and the gradient instability, it is common to add some epsilon values to the `Sigma` matrixes. 
We have tested the `torch.eye`[link](https://github.com/Michaelvll/DeepCCA/blob/2fa648e444fa6ae9135c52cceee1b653cc1d44d0/objectives.py#L37)
and it worked in most cases, but if your data has super little elements you can add another epsilon values to its elements, and as far as I know, it is a common way of increasing the stability.

I suggest testing the `torch.ones()` instead of the `torch.rand()`, because there is a chance that the rand element becomes near zero and after multiplying it with the `eps`, it becomes vanish.

Btw, you can use the `r1` or `r2` instead of `eps` as the the `equation (8)` of the main paper ( Galen Andrew, Raman Arora, Jeff Bilmes, Karen Livescu, ["Deep Canonical Correlation Analysis."](http://www.jmlr.org/proceedings/papers/v28/andrew13.pdf), ICML, 2013.) has explained.

**Comment by vandana-rajan** [comment]
Thanks. Glad to know that what I did is not wrong. It is working fine with my data and network now.
